File: backend/services/file_service.py
import os
import shutil
from typing import Tuple, Any
from fastapi import HTTPException

ALLOWED_EXTENSIONS = {".pdf", ".jpg", ".jpeg", ".png"}
ALLOWED_MIME_TYPES = {
    "application/pdf",
    "image/jpeg",
    "image/jpg",
    "image/png"
}
UPLOAD_DIRECTORY = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "uploads")

# Ensure uploads directory exists
os.makedirs(UPLOAD_DIRECTORY, exist_ok=True)

def validate_and_save_file(file: Any, prefix_id: str = "") -> Tuple[str, str]:
    """
    Validates uploaded file against allowed formats and stores it securely in uploads/.
    Returns (relative_file_path, absolute_file_path).
    """
    filename = file.filename or "upload.pdf"
    _, ext = os.path.splitext(filename)
    ext = ext.lower()

    if ext not in ALLOWED_EXTENSIONS:
        raise HTTPException(
            status_code=400,
            detail=f"Unsupported file format '{ext}'. Allowed formats: PDF, JPG, JPEG, PNG."
        )

    # Sanitize and create distinct filename
    clean_name = os.path.basename(filename).replace(" ", "_")
    if prefix_id:
        target_filename = f"{prefix_id}_{clean_name}"
    else:
        target_filename = f"{int(os.path.getmtime(UPLOAD_DIRECTORY) if os.path.exists(UPLOAD_DIRECTORY) else 1)}_{clean_name}"

    dest_path = os.path.join(UPLOAD_DIRECTORY, target_filename)

    with open(dest_path, "wb") as buffer:
        shutil.copyfileobj(file.file, buffer)

    relative_path = os.path.join("uploads", target_filename).replace("\\", "/")
    return relative_path, dest_path

File: backend/src/services/fileService.ts
import fs from 'fs';
import path from 'path';
import multer from 'multer';
import { BASE_DIR } from '../database/index.js';

export const ALLOWED_EXTENSIONS = new Set(['.pdf', '.jpg', '.jpeg', '.png', '.docx', '.doc', '.webp', '.bmp', '.tiff']);

function getUploadDir(): string {
  if (process.env.VERCEL || process.env.AWS_LAMBDA_FUNCTION_NAME) {
    const tmpUploadDir = path.join('/tmp', 'uploads');
    if (!fs.existsSync(tmpUploadDir)) {
      fs.mkdirSync(tmpUploadDir, { recursive: true });
    }
    return tmpUploadDir;
  }
  const uploadDir = path.join(BASE_DIR, 'uploads');
  if (!fs.existsSync(uploadDir)) {
    fs.mkdirSync(uploadDir, { recursive: true });
  }
  return uploadDir;
}

export const UPLOAD_DIRECTORY = getUploadDir();

// Multer in-memory storage for safe processing
export const uploadMiddleware = multer({
  storage: multer.memoryStorage(),
  limits: {
    fileSize: 25 * 1024 * 1024, // 25MB max
  },
  fileFilter: (_req, file, cb) => {
    const ext = path.extname(file.originalname).toLowerCase();
    if (ALLOWED_EXTENSIONS.has(ext)) {
      cb(null, true);
    } else {
      cb(new Error(`Unsupported file format '${ext}'. Allowed formats: PDF, JPG, JPEG, PNG, DOCX.`));
    }
  },
});

export function saveUploadedBuffer(buffer: Buffer, originalFilename: string, prefixId: string = ''): { relativePath: string; absolutePath: string } {
  const ext = path.extname(originalFilename).toLowerCase() || '.png';
  const cleanName = path.basename(originalFilename).replace(/\s+/g, '_');
  const targetFilename = prefixId ? `${prefixId}_${cleanName}` : `${Date.now()}_${cleanName}`;
  const destPath = path.join(UPLOAD_DIRECTORY, targetFilename);

  fs.writeFileSync(destPath, buffer);

  const relativePath = path.posix.join('uploads', targetFilename);
  return { relativePath, absolutePath: destPath };
}
